### PR TITLE
Fix floating point inaccuracies when digitizing (to fix bug #13745)

### PR DESCRIPTION
**Pull request by dgoedkoop**
Second try after #2434.

~~By putting the code in QgsMapToolCapture::nextPoint(), both the snapping function and the tracing function benefit from it.~~

I make this PR now so that anyone can take a look already (and @wonder-sk, so that you can see whether I understood your suggestion correctly), but it is not yet finished. For example:
- ~~Unit test?~~
- [x] Code for adding single points
- ~~Tracer~~

### CODE REVIEW

**Comment by wonder-sk** [comment]
Actually I meant something even simpler. The scenario in my head was:
1. get snapping match as before (no change) in map units
2. call a utility function to take the match (in map units) and return a new match in layer units
3. done :-)

The utility function would use the information already present in Match to fetch the original feature from the vector layer and replace the coordinates:
- if snapped to a vertex - just use the coords according to the vertex index
- if snapped to an edge - replace edge coords according to the vertex index + transform the intersection point
- if snapped to an intersection - just transform the intersection point

Does that make sense? In this scenario, there is no need for extra snap calls, just a bit of post-processing where necessary (e.g. QgsMapMouseEvent class may allow callers access the internal mSnapMatch to do the transform).

**Comment by dgoedkoop** [comment]
Ah I see, I didn't know that the Match struct contains all of that information. However, that would only work if the point comes directly from QgsMapMouseEvent, and not when adding a whole batch of points coming from QgsMapCanvasTracer (which, if I understand it correctly, internally merges the layers for snapping into one single graph in map coordinates, and snaps to that).

So, then, I think both variants are necessary.

**Comment by wonder-sk** [comment]
Correct, my approach does not handle tracing. I would prefer to address tracing separately - with a bit of work we could include the extra information also in tracer's graph and recover the exact source layer's coordinates similarly to snapping Match.

**Comment by dgoedkoop** [comment]
Ok, that would be even better! I've updated my code now.

The code for the tracer should probably be a bit different. A single point in the tracer graph can be a merge from points in different layers with different crs'es (again: if I understand it correctly!), so I think the MapToolCapture code should query the tracer with a method that has both a point (in map coordinates) and the desired crs as parameters. Does that make sense?

**Comment by wonder-sk** [comment]
> The code for the tracer should probably be a bit different. A single point in the tracer graph can be a merge from points in different layers with different crs'es (again: if I understand it correctly!), so I think the MapToolCapture code should query the tracer with a method that has both a point (in map coordinates) and the desired crs as parameters. Does that make sense?

Indeed - tracer's graph may contain vertices coming from various CRSes, and some vertices are just intersections (not having a source vertex in any geometry).

Thanks for the update - the code generally looks good!

File: src/gui/qgsmapmouseevent.h [hunk 1]
```diff
@@ -94,6 +94,12 @@ class GUI_EXPORT QgsMapMouseEvent : public QMouseEvent
     inline QgsPoint mapPoint() const { return mMapPoint; }
 
     /**
+      * Returns the matching data from the most recently snapped point.
+      * @return the snapping data structure
+      */
+    QgsPointLocator::Match mapPointMatch() const { return mSnapMatch; }
```
Comment (wonder-sk) [inline src/gui/qgsmapmouseevent.h]: Please also add a note that the method was added in QGIS 2.14 so plugin developers know the requirements if they want to use the function + add the method to .sip file to make it available in Python.

File: src/gui/qgsmaptoolcapture.cpp [hunk 2]
```diff
@@ -347,18 +347,42 @@ int QgsMapToolCapture::nextPoint( const QPoint &p, QgsPoint &layerPoint, QgsPoin
...
     return 2;
   }
 
-  int res;
   QgsPoint layerPoint;
-  res = nextPoint( point, layerPoint );
-  if ( res != 0 )
+  QgsVectorLayer* vlayer = qobject_cast<QgsVectorLayer *>( mCanvas->currentLayer() );
+  if ( match.isValid() && match.hasVertex() &&
+       ( match.layer()->crs() == vlayer->crs() ) )
```
Comment (wonder-sk) [inline src/gui/qgsmaptoolcapture.cpp]: When snapping to intersections is turned on, a match to an intersection is "vertex" match with no layer associated - we need to check for nullptr in match.layer()

**Comment by dgoedkoop** [comment]
Is a unit test necessary? I don't see other unit tests for the digitising tools.

Also, I have tried to understand more about how the tracer works. To me it looks like that it will be a bit more complicated to add some kind of 'source information' to the vertices of its internal graph. So maybe that should be done separately from this PR?

**Comment by wonder-sk** [comment]
> Is a unit test necessary? I don't see other unit tests for the digitising tools.

Unit tests for map tools would be nice, but currently they are not required - there is not groundwork yet in QGIS to support unit testing of map tools.

> So maybe that should be done separately from this PR?

That would be my preference, too.

**Comment by dgoedkoop** [comment]
Ok. Is there anything left to do for this PR, then?

**Comment by wonder-sk** [comment]
Thanks, the code looks good, just one last thing - would you mind putting that transform code into a protected function in QgsMapToolCapture? The same code is duplicated in QgsMapToolAddFeature - wrapping it into one function would de-duplicate the code and make it easier to understand what's going on there...

**Comment by dgoedkoop** [comment]
Done, thank you for the hint

**Comment by dgoedkoop** [comment]
Tomorrow I'd push a solution for the potential merge conflict (seems like a minor issue, but still want to test it). However, in order to save work, I have squashed all commits of this PR together before doing a rebase; so afterwards the discussion above will probably look a bit strange.

**Pull request merged**